File: project2/Human.java
import java.io.Serializable;

class Human extends Animal implements Serializable {
    private static final long serialVersionUID = 1L;
    private int dx = 0, dy = 0;
    int abilityCooldown = 0;
    int abilityDuration = 0;
    
    public Human(int x, int y, World w) {
        super(5, 4, x, y, w);
    }
    
    @Override
    public Animal clone(int x, int y) {
        return new Human(x, y, world);
    }
    
    public void setDirection(int dx, int dy) {
        this.dx = dx;
        this.dy = dy;
    }
    
    public void activateAbility() {
        if (abilityCooldown == 0) {
            abilityDuration = 5;
            abilityCooldown = 10;
            world.lastCollisionMessage = "Ability activated!";
            strength += 5;
        }
    }
    
    @Override
    public void action() {
        int nx = x + dx, ny = y + dy;
        dx = dy = 0;
        
        if (nx >= 0 && ny >= 0 && nx < 40 && ny < 20) {
            if (world.grid[ny][nx] != null && (nx != x || ny != y)) {
                collision(world.grid[ny][nx]);
            }
            if (alive) {
                world.grid[y][x] = null;
                x = nx;
                y = ny;
                world.grid[y][x] = this;
            }
        }
        
        if (abilityCooldown > 0) abilityCooldown--;
        if (abilityDuration > 0) {
            abilityDuration--;
            strength--;
        }
        
        age++;
    }
    
    @Override
    public char draw() {
        return 'H';
    }
}

File: project3/human.py
from animal import Animal

class Human(Animal):
    def __init__(self, x, y, world):
        super().__init__(5, 4, x, y, world)
        self.dx = 0
        self.dy = 0
        self.ability_cooldown = 0
        self.ability_duration = 0
    
    def clone(self, x, y):
        return Human(x, y, self.world)
    
    def set_direction(self, dx, dy):
        self.dx = dx
        self.dy = dy
    
    def activate_ability(self):
        if self.ability_cooldown == 0:
            self.ability_duration = 5
            self.ability_cooldown = 10
            self.world.last_collision_message = "Ability activated!"
            self.strength += 5
    
    def action(self):
        nx, ny = self.x + self.dx, self.y + self.dy
        self.dx, self.dy = 0, 0
        
        if 0 <= nx < self.WIDTH and 0 <= ny < self.HEIGHT:
            if self.world.grid[ny][nx] is not None and (nx != self.x or ny != self.y):
                self.collision(self.world.grid[ny][nx])
            
            if self.alive:
                self.world.grid[self.y][self.x] = None
                self.x, self.y = nx, ny
                self.world.grid[self.y][self.x] = self
        
        if self.ability_cooldown > 0:
            self.ability_cooldown -= 1
            
        if self.ability_duration > 0:
            self.ability_duration -= 1
            self.strength -= 1
        
        self.age += 1
    
    def draw(self):
        return 'H'
